src/openrct2/core/Http.WinHttp.cpp

```diff
@@ -0,0 +1,234 @@
+        {
+            case Method::GET:
+                return L"GET";
+            case Method::POST:
+                return L"POST";
+            case Method::PUT:
+                return L"PUT";
+            default:
+                throw std::runtime_error("Unsupported verb.");
+        }
+    }
+
+    static int32_t ReadStatusCode(HINTERNET hRequest)
+    {
+        wchar_t headerBuffer[32]{};
+        auto headerBufferLen = (DWORD)std::size(headerBuffer);
+        if (!WinHttpQueryHeaders(
+                hRequest, WINHTTP_QUERY_STATUS_CODE, L"StatusCode", headerBuffer, &headerBufferLen, WINHTTP_NO_HEADER_INDEX))
+        {
+            ThrowWin32Exception("WinHttpQueryHeaders");
+        }
+        auto statusCode = std::stoi(headerBuffer);
+        return statusCode < 0 || statusCode > 999 ? 0 : statusCode;
+    }
+
+    static std::map<std::string, std::string> ReadHeaders(HINTERNET hRequest)
+    {
+        DWORD dwSize{};
+        WinHttpQueryHeaders(
+            hRequest, WINHTTP_QUERY_RAW_HEADERS, WINHTTP_HEADER_NAME_BY_INDEX, NULL, &dwSize, WINHTTP_NO_HEADER_INDEX);
+
+        std::wstring buffer;
+        buffer.resize(dwSize);
+        if (!WinHttpQueryHeaders(
+                hRequest, WINHTTP_QUERY_RAW_HEADERS, WINHTTP_HEADER_NAME_BY_INDEX, buffer.data(), &dwSize,
+                WINHTTP_NO_HEADER_INDEX))
+            ThrowWin32Exception("WinHttpQueryHeaders");
+
+        std::map<std::string, std::string> headers;
+        std::wstring wKey, wValue;
+
+        constexpr int32_t STATE_EXPECT_KEY = 0;
+        constexpr int32_t STATE_EXPECT_WHITESPACE_VALUE = 1;
+        constexpr int32_t STATE_EXPECT_VALUE = 2;
+        int32_t state = STATE_EXPECT_KEY;
+        int32_t index = 0;
+        for (auto c : buffer)
+        {
+            if (c == '\0')
+            {
+                // Ignore first header as that is the HTTP version which
+                // we don't really count as a header.
+                if (index != 0 && wKey.size() != 0)
+                {
+                    auto key = String::ToUtf8(wKey);
+                    auto value = String::ToUtf8(wValue);
+                    headers[key] = value;
+                }
+                wKey.clear();
+                wValue.clear();
+                index++;
+                state = STATE_EXPECT_KEY;
+                continue;
+            }
+            if (state == STATE_EXPECT_KEY && c == ':')
+            {
+                state = STATE_EXPECT_WHITESPACE_VALUE;
+            }
+            else if (state == 1 && c == ' ')
+            {
+                state = STATE_EXPECT_VALUE;
+            }
+            else if (state == STATE_EXPECT_KEY)
+            {
+                wKey.push_back(c);
+            }
+            else
+            {
+                state = STATE_EXPECT_VALUE;
+                wValue.push_back(c);
+            }
+        }
+        return headers;
+    }
+
+    static std::string ReadBody(HINTERNET hRequest)
```

Comment:
I think it makes more sense to use std::vector<uint8_t> as a result since we are dealing with binary.

Reply:
I get your point but it would just been changing the http response struct used in curl code and everywhere else.

Reply:
I see, nevermind then.